sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystem.java

```diff
@@ -35,6 +36,29 @@
  * Clients should use {@link FileSystems} utility.
  */
 public abstract class FileSystem<ResourceIdT extends ResourceId> {
+  /**
+   * This is the entry point to convert users provided specs to {@link ResourceIdT ResourceIds}.
+   * Callers should use {@link #match} to resolve users specs ambiguities before
+   * calling other methods.
+   *
+   * <p>Implementation should handle the following ambiguities of a user provided spec:
+   * 1). spec could be a glob or a uri. {@link #match} should be able to tell and
+   *     choose efficient implementations.
+   * 2). spec does not end with a path delimiter, such as ‘/’, may refer to files or directories:
```

Comment:
I don't understand 2.

I do not believe that `file:/home/dir` is a legal URI. Is is a legal ResourceId?